refactor: introduce `mergeWithDefaults` and organize how default values for config options are set

sapphi-red opened the pull request:
### Description

Tentative PR to discuss how we should expose the values.

<!-- What is this PR solving? Write a clear description or reference the issues it solves (e.g. `fixes #123`). What other alternatives have you explored? Are there any parts you think require more attention from reviewers? -->

<!----------------------------------------------------------------------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md.
- Check that there isn't already a PR that solves the problem the same way. If you find a duplicate, please help us reviewing it.
- Update the corresponding documentation if needed.
- Include relevant tests that fail without this PR but pass with it.

Thank you for contributing to Vite!
----------------------------------------------------------------------->

sapphi-red committed:
wip: configDefaults

sapphi-red commented on packages/vite/src/node/build.ts:
```diff
@@ -401,17 +401,14 @@ export function resolveBuildEnvironmentOptions(
+    ...configDefaults.build,
     ...userBuildEnvironmentOptions,
     commonjsOptions: {
-      include: [/node_modules/],
-      extensions: ['.js', '.cjs'],
+      ...configDefaults.build.commonjsOptions,
       ...userBuildEnvironmentOptions.commonjsOptions,
     },
     dynamicImportVarsOptions: {
-      warnOnError: true,
-      exclude: [/node_modules/],
+      ...configDefaults.build.dynamicImportVarsOptions,
       ...userBuildEnvironmentOptions.dynamicImportVarsOptions,
     },
```
Probably needs a helper function that merges recursively but replaces arrays, slightly different from `mergeConfig`.

sapphi-red replied:
Added `mergeWithDefaults` function 👍

sapphi-red commented on packages/vite/src/node/config.ts:
```diff
@@ -589,6 +587,166 @@ export type ResolvedConfig = Readonly<
 >
 
+// inferred ones are omitted
+export const configDefaults = Object.freeze({
```
There's many inferred options. Is it fine to simply omit it?

bluwy replied:
I was initially thinking that we'd only export the defaults that're special, like hardcoded array of values. That way we solve the problem of users wanting to extend it rather than completely replacing it.

I'm not sure yet about exposing all the defaults in the object. It seems to be a lot to keep track of, and like you mention here, there's some that's inferred.

patak-dev replied:
I'm fine with omitting the inferred ones, maybe adding a comment about what how that inference works. My vote goes for the current implementation in the PR. I think once the "merge" helper is in, it is going to be easier to work with.

bluwy replied:
What kind of merge helper are you referring to? What I guess I'm concerned about with exposing tons of options this way is that people could unknowingly used it like `vite.build(configDefaults)` or use it to merge entire configs which could be wasteful or lead to incorrect merges (like arrays). Hence I was thinking to be a bit conservative here 🤔

patak-dev replied:
I was thinking on the `mergeWithDefaults` that Sapphi implemented. We could prevent `vite.build(configDefaults)` by checking the instance. But you're right that they may do something like: `vite.build({ ...configDefaults, foo: 'bar' })`.

I don't think that current users would ever try to do something like this, but depending how we document `configDefaults`, I see your point.

It feels a bit cumbersome to use if we expose a ton of variables like `CONFIG_DEFAULTS_RESOLVE_CONDITIONS`. Imagine having to import 3 or 4 of these.

Maybe we could make all the properties in the public `configDefaults` non-enumerable so they can not iterate over the keys and the user needs to access them one by one. We should get the same properties as the individual variables but with better typing and easier to import. We would still have a `internalConfigDefaults` that we can iterate to be used in `mergeWithDefaults`

sapphi-red replied:
I'm fine with trimming down the values for the exposed object. I think the object form makes it easier for users to find out the value for each options. But given that we already need to expose `mainFields` and `conditions` separately, maybe that advantage is gone.

bluwy replied:
`mergeWithDefaults` makes sense to me, I think we needed that elsewhere regardless of this PR too.

I guess given:
1. We couldn't expose all defaults as some are inferred.
2. Some defaults are based on the client / server consumer (`mainFields` and `conditions`), which Sapphi has now refactored as `DEFAULT_CLIENT_MAIN_FIELDS` etc (https://github.com/vitejs/vite/pull/18550#discussion_r1830293791)
3. And personally I think I'm missing the usecase where users would want to know the defaults via `configDefaults`. Isn't the problem we want to solve here that if someone wants to add a new condition on `resolve.conditions` but want to extend from Vite defaults, they can use this new export? What are the usecase for e.g. `configDefaults.base`?

I think I'm leaning more on simply exporting the constants like what Sapphi has now with `DEFAULT_CLIENT_MAIN_FIELDS`. We only export the ones that solves the usecase I mentioned in no3.

patak-dev replied:
Maybe it is a good idea to merge only the refactoring that makes things cleaner internally for now, and then we can discuss how to expose the defaults in another PR. Maybe we don't need to do this for Vite 6, people can still copy and past the defaults from the docs (it isn't the same though, that will fix the values, instead of adding an extra to the defaults that is what we have now but it may be ok)

sapphi-red replied:
I pushed a commit that removes the new exports so that we can separate the discussion about exposing the default values. 👍

sapphi-red commented on packages/vite/src/node/config.ts:
```diff
@@ -757,22 +915,22 @@ function resolveEnvironmentResolveOptions(
     consumer === 'client' || isSsrTargetWebworkerEnvironment
-      ? DEFAULT_CONDITIONS.filter((c) => c !== 'node')
-      : DEFAULT_CONDITIONS.filter((c) => c !== 'browser')
+      ? configDefaults.resolve.conditions.filter((c) => c !== 'node')
+      : configDefaults.resolve.conditions.filter((c) => c !== 'browser')
```
It might be confusing that `configDefaults.resolve.conditions` contains both `node` and `browser`.

patak-dev replied:
Maybe we need `serverConfigDefaults.resolve.conditions`? This is a hard one. I think as you say having both values would be confusing.

Or maybe we should omit this one. And treat it as inferred. We could expose some defaults in a special way?
`clientDefaults.resolve.conditions`, `serverDefaults.resolve.conditions`?

sapphi-red replied:
I removed `mainFields` and `conditions` from `configDefaults` and exposed the values separately. I think `clientDefaults` / `serverDefaults` might be confusing as it is also used for `ssr.target === 'webworker'`. If we remove all `isSsrTargetWebworkerEnvironment` in future, I think we can have `clientDefaults` / `serverDefaults` that contains all values that is applied conditionally by `consumer`.

sapphi-red committed:
refactor: define `mergeWithDefaults` and use it

sapphi-red committed:
refactor: use `mergeWithDefaults` in more places

sapphi-red committed:
refactor: use constants

sapphi-red committed:
chore: merge main

sapphi-red committed:
feat: remove mainFields and conditions from configDefaults and expose them separately

sapphi-red commented on packages/vite/src/node/index.ts:
```diff
@@ -7,6 +7,7 @@ export {
   loadConfigFromFile,
   resolveConfig,
   sortUserPlugins,
+  configDefaults,
```
I noticed that this only exposes the value for ESM entry point and does not expose the value for CJS entry point. Are we fine with that? To make that work, `configDefaults` and the values used in it needs to be moved to `constants.ts` and makes a bit difficult to read the code.

bluwy replied:
Yeah I think we'd have to expose it in CJS for consistency unfortunately, otherwise they could get a confusing error of it not working.

sapphi-red committed:
chore: don't expose anything

sapphi-red committed:
chore: merge main

patak-dev commented:
/ecosystem-ci run

patak-cat reviewed (dismissed):
this is awesome

sapphi-red committed:
refactor: make `resolvedConfig.preview` `ResolvedPreviewOptions`

sapphi-red committed:
fix: don't set default value of `server.host`

sapphi-red committed:
test: use toStrictEqual

sapphi-red committed:
fix: use clone build options for compat

sapphi-red committed:
refactor: extract `builderOptionsDefaults`

sapphi-red committed:
refactor: clone defaults

sapphi-red commented:
/ecosystem-ci run

sapphi-red commented:
I fixed unocss by a5e77f047bfb23b0e8a1ccd6a62c5f3e8f7f0264 and 3313e83685567d7d73c7e4a98ed35e3fab385dc2 and also fixed vite-setup-catalogue by fd4ccb9521db1157dbf6569eca7b2ce9bf9c3bb6.

The other fails that was not failing for latest scheduled are:

- marko: failing because we now inline SVGs in dev (#18581)
- qwik: will check later
- vite-environment-examples: type incompatibility caused by #18362

sapphi-red committed:
fix: don't set emptyOutDir

sapphi-red commented:
qwik should be fixed by https://github.com/vitejs/vite/pull/18550/commits/6a52f2fe3831ae2c2a4c0ca250157e0835143bbc
It seems storybook is also failing in main (https://discord.com/channels/804011606160703521/1304311018905800767/1304340637319041085)

sapphi-red commented:
/ecosystem-ci run

yannbf commented:
> qwik should be fixed by [6a52f2f](https://github.com/vitejs/vite/commit/6a52f2fe3831ae2c2a4c0ca250157e0835143bbc) It seems storybook is also failing in main ([discord.com/channels/804011606160703521/1304311018905800767/1304340637319041085](https://discord.com/channels/804011606160703521/1304311018905800767/1304340637319041085))

You already know but just for context for others:
Storybook started to fail because of an older commit, unrelated to this PR: https://github.com/vitejs/vite/commit/2ebe4b44430dd311028f72520ac977bb202ce50b#diff-89bae1df62862bb7f4a03d82a1e9cbf4ac6d0c042f21fbbacb0a2238bd050042R430

It only started to fail now because we updated our tests to also run our build command.

sapphi-red committed:
fix: don't set emptyOutDir

It was always overriding the value

sapphi-red committed:
refactor: remove inferred value from `buildEnvironmentOptionsDefaults`

sapphi-red commented:
/ecosystem-ci run

sapphi-red commented:
It should be fine now.

patak-cat reviewed (dismissed):
Let's merge this one @bluwy? [note: there is nothing new exposed now](https://github.com/vitejs/vite/pull/18550/files#r1831972475)

bluwy commented on packages/vite/src/node/utils.ts:
```diff
@@ -1069,6 +1070,74 @@ function backwardCompatibleWorkerPlugins(plugins: any) {
+  D extends Record<string, any>,
+  V extends Record<string, any>,
+>(defaults: D, values: V): MergeWithDefaultsResult<D, V> {
+  // NOTE: we need to clone the value here to avoid mutating the defaults
+  const merged: Record<string, any> = deepClone(defaults)
+  for (const key in values) {
```
Since `mergeWithDefaults` is being called recursively, when merging nested objects, wouldn't the object be cloned twice? Should we try to optimize this here? I think we'd have to make the cloning inside of `mergeWithDefaults`, and only clone for keys that are in `defaults` but not in `values`.

sapphi-red replied:
Good catch! Instead of doing that, I made the `deepClone` to be call only once before merging with defaults (ffb92fdf743026c98edd8830e0868878db36f060). This way, the properties of defaults that would be overridden with the values will be cloned unnecessarily compared to your suggestion, but I think that's negligible.

bluwy replied:
Works for me!

sapphi-red committed:
chore: merge main

sapphi-red committed:
perf: clone only once

sapphi-red committed:
fix: I thought I replaced it 😅

bluwy commented on packages/vite/src/node/utils.ts:
```diff
@@ -1069,6 +1070,82 @@ function backwardCompatibleWorkerPlugins(plugins: any) {
+  if (typeof value === 'function') {
+    return value as T
+  }
+  return structuredClone(value)
```
Since arrays, objects and functions are already handled above, what cases are covered here with the `structuredClone`? From a [quick test](https://perf.link/#eyJpZCI6InhtNWliaWk2ajN6IiwidGl0bGUiOiJzdHJ1Y3R1cmVkQ2xvbmUgaW1wYWN0IiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IFsuLi5BcnJheSgxMDAwKS5rZXlzKCldIiwidGVzdHMiOlt7Im5hbWUiOiJTaW1wbGUiLCJjb2RlIjoiZGF0YS5maW5kKHggPT4geCA9PT0gMTAwKSIsInJ1bnMiOlsxMDAwLDg3NDAwMCw2MTAwMCwxNTQwMDAsMTQ4NTAwMCw2MTMwMDAsMTA3NTAwMCwxMDAwLDE0MzAwMDAsMjIyMDAwMCwxMzQwMDAsMTM3MDAwLDEwMDAwLDUwMDAsNjkwMDAsMjE5MDAwLDMwMzAwMCwzMzIwMDAsMjQxMDAwLDIwMDAsMjU4MDAwLDU5ODAwMCwzNzAwMDAsNTAwMCw1NDAwMCwxMzgwMDAsNDgwMDAsNDAwMDAsMTQ5MDAwLDY5MzAwMCwzMTAwMCw2NDEwMDAsMTM1MDAwLDM0NzAwMCwxNDEwMDAsMTcwMDAsMTEwMDAwLDE3MzAwMCw1MzcwMDAsNzAwMCwxMzAwMCwyMjcwMDAsMTQxMDAwLDEzNDAwMCwxNTgwMDAsMTM1MDAwLDEzMjAwMCwxNTIwMDAsMjUwMDAsODAwMDAsNjcwMDAsMTIwNDAwMCwyMDM5MDAwLDIzMjAwMCwxMTAwMDAsNTQwMDAsNjcwMDAsMTg5MDAwLDkwMjAwMCwxNjYwMDAsNTgwMDAsNDYwMDAsMTAyMDAwLDQyNjAwMCwyNzAwMCwxOTgwMDAsMTYyMDAwLDIxMjkwMDAsMTQwMDAsMjMzMDAwLDExNzAwMCwyNDcwMDAsNTEwMDAsMTEyMDAwLDIwMzAwMCwxNTEwMDAsNTE0MDAwLDEwODAwMCw4MTAwMCw1NDUwMDAsMTUyMDAwLDQ3MDAwLDMxMDAwMCw1MDAwLDEzMjAwMCwyMzAwMDAsMTczMDAwLDUxODAwMCw4MjAwMCwyMDAwLDE5MDAwMCwzNjIwMDAsMjA0MDAwLDI1ODAwMCwyMDAwLDEyNzAwMCwxNDQwMDAsMTAzMDAwLDU2MjAwMCwzMTYwMDBdLCJvcHMiOjI5NTI4MH0seyJuYW1lIjoic3R1Y3R1cmVkQ2xvbmUiLCJjb2RlIjoiZGF0YS5maW5kKHggPT4gc3RydWN0dXJlZENsb25lKHgpID09PSAxMDApIiwicnVucyI6WzIwMDAsMzAwMCwyNTAwMCwxNjAwMCwyNDAwMCwyNjAwMCwyNDAwMCwzMTAwMCwyMTAwMCwzODAwMCw1MDAwLDQwMDAsMTAwMCwxMDAwLDMwMDAsODAwMCwxMjAwMCw2MDAwLDQwMDAsMzAwMCwxMTAwMCw1MDAwLDEyMDAwLDYwMDAsNjAwMCw1MDAwLDcwMDAsNjAwMCw5MDAwLDE4MDAwLDMwMDAsNTAwMCwxMjAwMCwxMDAwMCwxMjAwMCwyMDAwLDYwMDAsNzAwMCwxNjAwMCwyMDAwLDUwMDAsMTAwMDAsNDAwMCw1MDAwLDEyMDAwLDE5MDAwLDcwMDAsNzAwMCw4MDAwLDExMDAwLDUwMDAsMjQwMDAsMzYwMDAsMTAwMDAsMTMwMDAsNDAwMCw0MDAwLDgwMDAsNDAwMCw4MDAwLDcwMDAsMTAwMDAsNDAwMCwxNTAwMCw4MDAwLDEwMDAwLDUwMDAsMjcwMDAsMTQwMDAsNjAwMCw2MDAwLDYwMDAsMjAwMCw1MDAwLDgwMDAsNjAwMCwxNDAwMCwxMDAwLDcwMDAsOTAwMCw3MDAwLDgwMDAsODAwMCwzMDAwLDYwMDAsMTEwMDAsMTIwMDAsOTAwMCwxMDAwMCwxMDAwLDcwMDAsNzAwMCw5MDAwLDkwMDAsMTQwMDAsMzAwMCw3MDAwLDgwMDAsOTAwMCw4MDAwXSwib3BzIjo5NDcwfV0sInVwZGF0ZWQiOiIyMDI0LTExLTEyVDA2OjQzOjA1LjAyOFoifQ%3D%3D), `structuredClone` seems to have a big impact when used on basic primitives.

sapphi-red replied:
It was mainly for `RegExp`. I put the `structuredClone` in a condition 👍 (38162a0a1603da5de06d683d97a1b175289caeeb, e4364e59fbd760c2f7f6e63a2cc137720cd2690d).

bluwy replied:
I don't think it's likely for someone to mutate a `RegExp`, but I don't mind having the code for it either ways.

sapphi-red committed:
perf: avoid `structuredClone`

sapphi-red committed:
fix: skip null and RegExp

bluwy reviewed (approved):
Looks great! Thanks for updating based on the suggestions.
